### DISCUSSION TITLE
Is there a way to access prefetched data on the server?

### DISCUSSION THREAD

**Discussion by brattonross** [trpc/trpc#1496] [Q&A]
Hi, thank you for your hard work on this library, it is making my development experience a lot more pleasant so far :)

I am migrating an existing Next.js app, built with react-query, to tRPC. In some of my pages I am prefetching data inside of `getServerSideProps`, and then accessing that data via `queryClient.getQueryData` so that I can execute some logic based on the result. In my case I am using the result to determine which locale namespaces need to be prefetched.

Is there an equivalent in tRPC, or some way to get hold of the QueryClient that tRPC has used so that I can get the data from it?

**Comment by aymenhmaidiwastaken** [comment]
Yes — use the `createCaller` API to call your tRPC procedures directly on the server without an HTTP round-trip.

```ts
import { appRouter } from '~/server/routers/_app';
import { createCallerFactory } from '~/server/trpc';

const createCaller = createCallerFactory(appRouter);

export async function getServerSideProps(ctx) {
  const caller = createCaller({
    // pass your context (session, db, etc.)
    session: await getSession(ctx.req),
  });

  // Call procedures directly — no HTTP, no serialization
  const posts = await caller.post.list();
  const user = await caller.user.me();

  return {
    props: { posts, user },
  };
}
```

With `createCaller`, the prefetched data is available synchronously in your component props. No need for React Query's cache on the server side.

If you're using React Query with SSR and want to hydrate the cache:

```ts
import { createServerSideHelpers } from '@trpc/react-query/server';

export async function getServerSideProps() {
  const helpers = createServerSideHelpers({
    router: appRouter,
    ctx: { session: ... },
  });

  await helpers.post.list.prefetch();

  return {
    props: {
      trpcState: helpers.dehydrate(),
    },
  };
}
```

This prefetches on the server and hydrates the React Query cache on the client — so `useQuery` picks up the data instantly without refetching.
